runtime/executor/src/main/java/edu/snu/nemo/runtime/executor/data/BlockTransferConnectionQueue.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2018 Seoul National University
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package edu.snu.nemo.runtime.executor.data;
+
+import edu.snu.nemo.conf.JobConf;
+import org.apache.reef.tang.annotations.Parameter;
+
+import javax.inject.Inject;
+import java.util.ArrayDeque;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A class to restrict parallel connection per runtime edge.
+ */
+public final class BlockTransferConnectionQueue {
+  private static final Object FUTURE_OBJECT = new Object();
+  private final Map<String, Integer> runtimeEdgeIdToNumOutstandingConnections = new ConcurrentHashMap<>();
+  private final Map<String, Queue<CompletableFuture<Object>>> runtimeEdgeIdToPendingConnections
+      = new ConcurrentHashMap<>();
```

Comment:
How about `HashMap`, since these two variables are only used in `synchronized` methods?

[This](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html) warns on the correctness of methods like `size` or `isEmpty` under `ConcurrentHashMap`:

> Bear in mind that the results of aggregate status methods including size, isEmpty, and containsValue are typically useful only when a map is not undergoing concurrent updates in other threads. Otherwise the results of these methods reflect transient states that may be adequate for monitoring or estimation purposes, but not for program control.


Also, [this](https://stackoverflow.com/questions/25998536/does-a-concurrenthashmap-need-wrapped-in-a-synchronized-block) notes that the virtue of `ConcurrentHashMap` is for better performance via locking only *part* of the table, whose benefit would be lost because it's used in `synchronized` methods anyway.